Swift 3.0

jakecraige opened the pull request:
Built off @sharplet's open PR for 2.3 support

Do the minimum set of changes to support Swift 3.

There are other improvements we can make like:
- using `Any` instead of `AnyObject` (JSONSerialization returns this now)
- Adopt a more Swify API by not ignoring so many first args.

Open question(s):
- Now that we have `URLRequest`, should we still support `NSURLRequest` and `NSMutableURLRequest`?

jakecraige committed:
Remove carthage installation from setup/update

As of Xcode 8, Carthage now recommends installing via the `.pkg` file instead of
Homebrew.

jakecraige committed:
Upgrade to Swift 3 dependencies

Upgrade to Swift 3 compatible dependencies, include Runes as an explicit
dependency to work around a current issue with Argo: https://github.com/thoughtbot/Argo/issues/412

jakecraige committed:
Convert existing API to Swift 3

Do the minimum set of changes to support Swift 3.

There are other improvments we can make like:
  - using `Any` instead of `AnyObject` (JSONSerialization returns this now)
  - Adopt a more Swify API by not ignoring so many first args.

Open question(s):

- Now that we have `URLRequest`, should we still support `NSURLRequest` and `NSMutableURLRequest`?

Eke commented:
Hello, any updates on this PR?

jakecraige commented:
@Eke We need to wait for #92 before merging this, but in the meantime you should be able to reference this branch `swift-3.0` in your dependency manager to use this version

Eke commented:
@jakecraige Thanks for info !

sharplet commented on .gitmodules:
```diff
@@ -1,12 +1,15 @@
 [submodule "Carthage/Checkouts/Result"]
 	path = Carthage/Checkouts/Result
-	url = https://github.com/antitypical/Result.git
+	url = https://0243d50631cefef6b22c779f29d6414b513d1654@github.com/antitypical/Result.git
```
This seems a little weird to me, a sha as a username in the URL?

sharplet commented on Source/Extensions/NSError.swift:
```diff
@@ -5,33 +5,33 @@ private let swishDomain = "com.thoughtbot.swish"
 }
 
-private func userInfoFor(function: String, _ file: String, _ line: Int) -> [String: AnyObject] {
+private func userInfoFor(_ function: String, _ file: String, _ line: Int) -> [String: AnyObject] {
```
I think we can go ahead and use `Any` here, should be able to remove most of the explicit type annotations.

sharplet commented on Source/Extensions/NSURLRequest.swift:
```diff
@@ -1,25 +1,18 @@
-  override var jsonPayload: AnyObject {
     get {
-      return super.jsonPayload
+      let json = httpBody.flatMap { try? JSONSerialization.jsonObject(with: $0, options: JSONSerialization.ReadingOptions(rawValue: 0)) }
```
I think `options:` gets a default value of `0`, so you should be able to remove that parameter completely.

sharplet commented on Source/Models/JSONDeserializer.swift:
```diff
@@ -2,33 +2,32 @@ import Argo
-      try NSJSONSerialization
-        .JSONObjectWithData(d, options: NSJSONReadingOptions(rawValue: 0))
-    )
+      attempt: { try JSONSerialization.jsonObject(with: d, options: JSONSerialization.ReadingOptions(rawValue: 0)) }
```
Same here about removing `options:` in favour of the default value.

sharplet commented on Source/Models/NetworkRequestPerformer.swift:
```diff
@@ -2,23 +2,23 @@ import Foundation
+  fileprivate let session: URLSession
 
-  public init(session: NSURLSession = NSURLSession.sharedSession()) {
+  public init(session: URLSession = URLSession.shared) {
```
Curious: would `.shared` work in this context?

sharplet commented on Source/Models/NetworkRequestPerformer.swift:
```diff
@@ -2,23 +2,23 @@ import Foundation
-  func performRequest(request: NSURLRequest, completionHandler: Result<HTTPResponse, SwishError> -> Void) -> NSURLSessionDataTask {
-    let task = session.dataTaskWithRequest(request) { data, response, error in
+  func performRequest(_ request: URLRequest, completionHandler: @escaping (Result<HTTPResponse, SwishError>) -> Void) -> URLSessionDataTask {
+    let task = session.dataTask(with: request, completionHandler: { data, response, error in
```
Trailing closure here?

sharplet commented on Source/Models/Scheduler.swift:
```diff
@@ -1,9 +1,9 @@
-public typealias Scheduler = ((() -> Void) -> Void)
+public typealias Scheduler = ((@escaping () -> Void) -> Void)
```
I feel like this annotation isn't in the right place. The closure that gets passed _into_ the scheduler shouldn't need to be escaping, right?

tonyd256 replied:
I'm not sure what @jakecraige's intention was here.

sharplet replied:
My bad, the closure may be scheduled asynchronously, so it makes sense for it to be `@escaping` here.

sharplet commented on Tests/Fakes/FakeRequestPerformer.swift:
```diff
@@ -2,56 +2,56 @@ import Foundation
-    .dataWithJSONObject(j, options: NSJSONWritingOptions(rawValue: 0))
+func serializeJSON(_ j: [String: AnyObject]) -> Data? {
+  return try? JSONSerialization
+    .data(withJSONObject: j, options: JSONSerialization.WritingOptions(rawValue: 0))
```
Default options here

sharplet commented on Tests/Tests/APIClientSpec.swift:
```diff
@@ -11,11 +11,11 @@ class APIClientSpec: QuickSpec {
             let client = APIClient(requestPerformer: performer)
 
-            client.performRequest(request) { _ in }
+            _ = client.performRequest(request) { _ in }
```
I think `performRequest()` should be annotated with `@discardableResult`.

sharplet reviewed (commented):
I say let's go ahead and switch to using `Any` instead of `AnyObject`. That should really tighten things up!

I think we can approach public API changes (like the first argument thing you mentioned) in a separate PR.

I think we should _not_ support `NSURLRequest` and `NSMutableURLRequest`.

Commit:
Update dependencies to Swift 3 versions

Commit:
Use Any instead of AnyObject

Commit:
Fetch tags for submodules

Commit:
maybe

Commit:
I LOVE SUBMODULES

Commit:
Why even bother

sharplet commented on Source/Models/JSONDeserializer.swift:
```diff
@@ -15,19 +15,19 @@ struct JSONDeserializer: Deserializer {
-      attempt: { try JSONSerialization.jsonObject(with: d, options: JSONSerialization.ReadingOptions(rawValue: 0)) }
-    ).map { $0 as AnyObject }
+      attempt: { try JSONSerialization.jsonObject(with: d) }
+    ).map { $0 }
```
Maybe this map isn't necessary anymore?

tonyd256 replied:
oh yep forgot to remove that

sharplet commented on Source/Models/APIClient.swift:
```diff
@@ -3,19 +3,21 @@ import Argo
-  public func performRequest<T: Request>(request: T, completionHandler: Result<T.ResponseObject, SwishError> -> Void) -> NSURLSessionDataTask {
+  @discardableResult
+  public func performRequest<T: Request>(_ request: T, completionHandler: @escaping (Result<T.ResponseObject, SwishError>) -> Void) -> URLSessionDataTask {
+
```
Extra newline here

sharplet commented on Source/Models/JSONDeserializer.swift:
```diff
@@ -2,33 +2,32 @@ import Argo
-    guard let d = data where d.length > 0 else {
-      return .Success(NSNull())
+  func parseJSON(_ data: Data?) -> Result<Any, NSError> {
+    guard let d = data , d.count > 0 else {
```
Extra space before the comma

sharplet commented on Source/Protocols/Client.swift:
```diff
@@ -3,5 +3,6 @@ import Argo
 public protocol Client {
-  func performRequest<T: Request>(request: T, completionHandler: Result<T.ResponseObject, SwishError> -> ()) -> NSURLSessionDataTask
+  @discardableResult
+  func performRequest<T: Request>(_ request: T, completionHandler: @escaping (Result<T.ResponseObject, SwishError>) -> ()) -> URLSessionDataTask
```
Thoughts on renaming this to `perform()`?

tonyd256 replied:
ya, seems like that is a good Swift 3 api change

sharplet commented on Source/Protocols/RequestPerformer.swift:
```diff
@@ -2,5 +2,6 @@ import Foundation
 public protocol RequestPerformer {
-  func performRequest(request: NSURLRequest, completionHandler: Result<HTTPResponse, SwishError> -> Void)-> NSURLSessionDataTask
+  @discardableResult
+  func performRequest(_ request: URLRequest, completionHandler: @escaping (Result<HTTPResponse, SwishError>) -> Void)-> URLSessionDataTask
```
Same here about renaming to `perform()`

Eke commented:
Hello, can you please update `podspec` dependencies for `Argo` and `Result` ?

eliperkins commented on circle.yml:
```diff
@@ -4,9 +4,7 @@ machine:
-    - git submodule init
-    - git submodule foreach --recursive 'git fetch --tags'
-    - git submodule update --recursive
+    - bin/setup
```
I wish it were possible to fav a commit. Because I would fav this series of submodule commits for sure.

Commit:
PR updates and Swiften it up

Commit:
Add back the _

ren commented:
Hey folks, any update on this getting merged? Is there any work which could be contributed to this PR to get it closer to ready?

sharplet committed:
Remove unnecessary casting

sharplet committed:
Don't conform SwishError to Equatable

Error equatability doesn't compose well with underlying errors that are
exposed to Swift as `Error` rather than `NSError`. Swift's error
handling model is based around pattern matching, so we should use that
instead of equatability, which is not guaranteed.

sharplet committed:
Remove NSError in favour of Swift.Error and Result.AnyError

Raises the Result requirement to 3.1, which introduced `AnyError`.

sharplet committed:
Fix CustomNSError implementation on Swift 3.0.0

Since Swift 3.0.1, `CustomNSError` provides default implementations of
`errorCode` and `errorUserInfo`. Unfortunately, in order to support
Swift 3.0.0, we need to add a custom implementation. Further, it's not
possible to specify a patch version in `#if swift()`, so in order to use
the default implementation we have to target at least Swift 3.1 (even
though the default implementation is available in Swift 3.0.1).

sharplet commented:
@jakecraige @tonyd256 I've pushed a few more commits removing some unnecessary casts and updating our error handling to use the new `LocalizedError` and `CustomNSError` protocols. I think this should be about ready to merge!

jakecraige commented on Source/Extensions/Result.swift:
```diff
@@ -2,12 +2,12 @@ import Argo
 
 public extension Result {
-  static func fromDecoded<T>(decoded: Decoded<T>) -> Result<T, SwishError> {
+  static func fromDecoded<T>(_ decoded: Decoded<T>) -> Result<T, SwishError> {
```
Since we're in Swift 3 now, what about changing this to `static func from<T>(decoded: ....`?

sharplet replied:
Good idea. I refactored it to be an initailiser in c5e4917.

jakecraige commented on Source/Models/APIClient.swift:
```diff
@@ -29,12 +30,12 @@ extension APIClient: Client {
 
 private extension APIClient {
-  func validateResponse(httpResponse: HTTPResponse) -> Result<NSData?, SwishError> {
+  func validateResponse(_ httpResponse: HTTPResponse) -> Result<Data?, SwishError> {
```
Same here about Swift 3 API guidelines

sharplet replied:
Works for me.

jakecraige commented on Source/Models/SwishError.swift:
```diff
@@ -2,47 +2,69 @@ import Foundation
+    case .deserializationError:   return 2
+    case .parseError:             return 3
+    case .serverError:            return 4
+    case .urlSessionError:        return 5
```
I don't think we have a guideline for Swift, but generally we prefer not to align things like these `returns` because they can cause a lot of diff noise if we ever change the name of the longest one or add a longer one and all the indenting has to be updated

sharplet replied:
Removing the aligment felt pretty messy and hard to scan for me. I ended up putting the returns statements on separate lines instead.

jakecraige commented on Source/Models/SwishError.swift:
```diff
@@ -2,47 +2,69 @@ import Foundation
+    var userInfo: [String: Any] = [:]
+    if let errorDescription = errorDescription {
+      userInfo[NSLocalizedDescriptionKey] = errorDescription
+    }
```
I think you can remove this unwrapping and set it explicitly because [`[]` accepts an optional](https://developer.apple.com/reference/swift/dictionary/1540848-subscript)?

sharplet replied:
Works for me.

jakecraige commented on bin/setup:
```diff
@@ -1,4 +1,3 @@
 #!/usr/bin/env bash
 
-brew install carthage 2> /dev/null
```
👏

sharplet committed:
Replace Result.fromDecoded with an initializer

sharplet committed:
Add fix-its for renamed methods

sharplet committed:
Rename validateResponse() to validate()

jakecraige commented:
👍  (can't approve through review since GH says this is "my" PR)
